File: src/main/java/io/gittul/domain/post/PostSocialController.java
package io.gittul.domain.post;

import io.gittul.domain.bookmark.entity.BookmarkService;
import io.gittul.domain.like.LikeService;
import io.gittul.domain.post.dto.PostFeedResponse;
import io.gittul.domain.user.entity.User;
import io.gittul.global.page.PageQuery;
import io.gittul.infra.auth.aop.Authenticated;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/posts")
@RequiredArgsConstructor
public class PostSocialController {

    private final PostService postService;
    private final LikeService likeService;
    private final BookmarkService bookmarkService;

    @GetMapping("/following")
    public List<PostFeedResponse> getFollowingPosts(@Authenticated User user,
                                                    PageQuery page) {
        return postService.getFollowingPosts(user, page.toRequest());
    }

    @GetMapping("/bookmark")
    public List<PostFeedResponse> getBookmarkedPosts(@Authenticated User user,
                                                     PageQuery page) {
        return postService.getBookmarkedPosts(user, page.toRequest());
    }

    @PostMapping("/{id}/like")
    public void likePost(@Authenticated User user, @PathVariable Long id) {
        likeService.likePost(user, id);
    }

    @PostMapping("/{id}/unlike")
    public void unLikePost(@Authenticated User user, @PathVariable Long id) {
        likeService.unLikePost(user, id);
    }

    @PostMapping("/{id}/bookmark")
    public void addBookmark(@Authenticated User user, @PathVariable Long id) {
        bookmarkService.addBookmark(user, id);
    }

    @PostMapping("/{id}/unbookmark")
    public void removeBookmark(@Authenticated User user, @PathVariable Long id) {
        bookmarkService.removeBookmark(user, id);
    }
}

File: app/src/main/kotlin/io/gittul/app/domain/post/PostSocialController.kt
package io.gittul.app.domain.post

import io.gittul.app.domain.bookmark.BookmarkService
import io.gittul.app.domain.like.LikeService
import io.gittul.app.domain.post.dto.PostFeedResponse
import io.gittul.app.infra.auth.aop.AccessGuard
import io.gittul.app.infra.auth.aop.AuthContext
import io.gittul.core.global.page.PageQuery
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@AccessGuard
@RestController
@RequestMapping("/posts")
class PostSocialController(
    private val postService: PostService,
    private val likeService: LikeService,
    private val bookmarkService: BookmarkService
) {


    @GetMapping("/following")
    fun getFollowingPosts(
        page: PageQuery
    ): List<PostFeedResponse> {
        val user = AuthContext.getUser()
        return postService.getFollowingPosts(user, page.toRequest())
    }

    @GetMapping("/bookmark")
    fun getBookmarkedPosts(
        page: PageQuery
    ): List<PostFeedResponse> {
        val user = AuthContext.getUser()
        return postService.getBookmarkedPosts(user, page.toRequest())
    }

    @PostMapping("/{id}/like")
    fun likePost(@PathVariable id: Long) {
        val user = AuthContext.getUser()
        likeService.likePost(user, id)
    }

    @PostMapping("/{id}/unlike")
    fun unLikePost(@PathVariable id: Long) {
        val user = AuthContext.getUser()
        likeService.unLikePost(user, id)
    }

    @PostMapping("/{id}/bookmark")
    fun addBookmark(@PathVariable id: Long) {
        val user = AuthContext.getUser()
        bookmarkService.addBookmark(user, id)
    }

    @PostMapping("/{id}/unbookmark")
    fun removeBookmark(@PathVariable id: Long) {
        val user = AuthContext.getUser()
        bookmarkService.removeBookmark(user, id)
    }
}
